kouta-backend/src/main/scala/fi/oph/kouta/validation/Validations.scala

```diff
@@ -337,6 +400,100 @@ object Validations extends Logging {
+      validationContext: ValidationContext,
+      errorMessage: ErrorMessage
+  ): IsValid = {
+    val queryResult = if (validationContext.isKoodistoServiceOk()) queryMethod(koodiUri) else queryFailed
+    validationContext.updateKoodistoServiceStatusByQueryStatus(queryResult)
+    assertExternalQueryResult(
+      queryResult,
+      path,
+      errorMessage,
+      koodistoServiceFailureMsg
+    )
+  }
+
+  def assertKoulutustyyppiQueryResult(
+      koulutusKoodiUri: String,
+      koulutusTyypit: Seq[String],
+      koulutusKoodiClient: KoulutusKoodiClient,
+      path: String,
+      validationContext: ValidationContext,
+      errorMessage: ErrorMessage,
+      koodistoServiceFailureMessage: ErrorMessage = koodistoServiceFailureMsg
+  ): IsValid = {
+    val queryResult =
+      if (validationContext.isKoodistoServiceOk())
+        koulutusKoodiClient.koulutusKoodiUriOfKoulutustyypitExist(koulutusTyypit, koulutusKoodiUri)
+      else queryFailed
+    validationContext.updateKoodistoServiceStatusByQueryStatus(queryResult)
+    assertExternalQueryResult(
+      queryResult,
+      path,
+      errorMessage,
+      koodistoServiceFailureMessage
+    )
+  }
+
+  def assertKoulutuskoodiQueryResult(
+      koulutusKoodiUri: String,
+      koulutusKoodiFilter: Seq[String],
+      koulutusKoodiClient: KoulutusKoodiClient,
+      path: String,
+      validationContext: ValidationContext,
+      errorMessage: ErrorMessage,
+      koodistoServiceFailureMessage: ErrorMessage = koodistoServiceFailureMsg
+  ): IsValid = {
+    val queryResult =
+      if (validationContext.isKoodistoServiceOk())
+        koulutusKoodiClient.koulutusKoodiUriExists(koulutusKoodiFilter, koulutusKoodiUri)
+      else queryFailed
+    validationContext.updateKoodistoServiceStatusByQueryStatus(queryResult)
+    assertExternalQueryResult(
+      queryResult,
+      path,
+      errorMessage,
+      koodistoServiceFailureMessage
+    )
+  }
+
+  def assertAtaruQueryResult(
+      ataruId: UUID,
+      hakemusPalveluClient: HakemusPalveluClient,
+      path: String,
+      validationContext: ValidationContext,
+      errorMessage: ErrorMessage
+  ): IsValid = {
+    val queryResult =
+      if (validationContext.isAtaruServiceOk()) hakemusPalveluClient.isExistingAtaruId(ataruId) else queryFailed
+    validationContext.updateAtaruServiceStatusByQueryStatus(queryResult)
+    assertExternalQueryResult(
+      queryResult,
+      path,
+      errorMessage,
+      ataruServiceFailureMsg
+    )
+  }
+
+  def assertExternalQueryResult(
+      externalQueryResult: ExternalQueryResult,
+      path: String,
+      errorMessage: ErrorMessage,
+      externalServiceFailureMessage: ErrorMessage
+  ): IsValid = {
+    // Pattern matching ei toimi ExternaQueryResultille
+    if (externalQueryResult == itemFound)
+      NoErrors
+    else if (externalQueryResult == queryFailed) {
+      error(path, externalServiceFailureMessage)
+    } else
+      error(path, errorMessage)
+  }
```

Comment:
Eikö tää toimi, missaanko jotain?
```suggestion
    externalQueryResult match {
      case e if e == itemFound => NoErrors
      case e if e == queryFailed => error(path, externalServiceFailureMessage)
      case _ => error(path, errorMessage)
    }
```